src/mongo/client/dbclient.cpp

```diff
@@ -1471,6 +1471,7 @@ list<string> DBClientWithCommands::getDatabaseNames() {
 
 list<string> DBClientWithCommands::getCollectionNames(const string& db, const BSONObj& filter) {
     auto_ptr<DBClientCursor> infos = enumerateCollections(db, filter);
+    uassert(0, "socket error for mapping query", infos.get());
```

Comment:
I also suspect that DBClientWithCommands::exists is similarly afflicted. Can you just check all callers of enumerateCollections?